extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -0,0 +1,357 @@
+   * parallel with other operations.
+   *
+   * @param {string} signingServiceName
+   */
+  loadKeyset(signingServiceName) {
+    if (this.signers_ && !this.signers_[signingServiceName]) {
+      const keys = {};
+      const promise = this.fetchAndAddKeys_(keys, signingServiceName, null);
+      this.signers_[signingServiceName] = {promise, keys};
+    }
+  }
+
+  /**
+   * Verifies that `signature` is the correct cryptographic signature for
+   * `creative`, with the public key from the named signing service identified
+   * by `keypairId`.
+   *
+   * As a precondition, `loadKeyset(signingServiceName)` must have already been
+   * called.
+   *
+   * If the keyset for the named signing service was imported successfully but
+   * did not include a key for `keypairId`, this may be the result of a stale
+   * browser cache. To work around this, `keypairId` is added to the public key
+   * endpoint URL as a query parameter and the keyset is re-fetched. Other kinds
+   * of failures, including network connectivity failures, are not retried.
+   *
+   * @param {string} signingServiceName
+   * @param {string} keypairId
+   * @param {!Uint8Array} signature
+   * @param {!ArrayBuffer} creative
+   * @return {!Promise<?VerificationFailure>} resolves to `null` on success, or,
+   *     on failure, to a `VerificationFailure` indicating the cause
+   */
+  verify(signingServiceName, keypairId, signature, creative) {
+    if (this.signers_) {
+      const signer = this.signers_[signingServiceName];
+      return signer.promise.then(success => {
+        if (success) {
+          const keyPromise = signer.keys[keypairId];
+          if (keyPromise === undefined) {
+            // We don't have this key, but maybe the cache is stale; try
+            // cachebusting.
+            signer.promise = this.fetchAndAddKeys_(
+                                     signer.keys, signingServiceName, keypairId)
+                                 .then(success => {
+                                   if (signer.keys[keypairId] === undefined) {
+                                     // We still don't have this key; make sure
+                                     // we never try again.
+                                     signer.keys[keypairId] = null;
+                                   }
+                                   return success;
+                                 });
+            // This "recursive" call can recur at most once.
+            return this.verify(
+                signingServiceName, keypairId, signature, creative);
+          } else if (keyPromise === null) {
+            // We tried cachebusting and still don't have this key.
+            return VerificationFailure.KEY_NOT_FOUND;
+          } else {
+            return keyPromise.then(key => {
+              if (key) {
+                return cryptoFor(this.win_)
+                    .verifyPkcs(key, signature, creative)
+                    .then(
+                        result => {
```

Comment:
This is because there are two callbacks here, so the rectangle rule requires that they both be on separate lines.